tests/subsys/bluetooth/mesh/sensor_subsys_new/src/main.c

```diff
@@ -0,0 +1,984 @@
+	zassert_equal(status, expected_status);
+	if (expected_micro == 0 || exact) {
+		zassert_equal(micro, expected_micro);
+	} else {
+		double error = (double)micro / (double)expected_micro;
+
+		zassert_between_inclusive(error, 0.99999, 1.00001);
+	}
+}
+
+static void check_from_float(const struct bt_mesh_sensor_format *format,
+			     float f, int expected_err,
+			     uint32_t expected_raw, bool exact)
+{
+	struct bt_mesh_sensor_value val = { 0 };
+	int err = bt_mesh_sensor_value_from_float(format, f, &val);
+
+	zassert_equal(err, expected_err);
+	zassert_equal(val.format, format);
+
+	uint32_t raw = sys_get_le32(val.raw);
+
+	double error;
+
+	if (expected_raw == 0 || exact) {
+		zassert_equal(raw, expected_raw);
+	} else if (f < 0) {
+		int32_t expected_raw_signed = *(int32_t *)(&expected_raw);
+		int32_t raw_signed = *(int32_t *)(&raw);
+
+		error = (double)raw_signed / expected_raw_signed;
+		zassert_between_inclusive(error, 0.99999, 1.00001);
+	} else {
+		error = (double)raw / expected_raw;
+		zassert_between_inclusive(error, 0.99999, 1.00001);
+	}
+}
+
+static void check_to_float(const struct bt_mesh_sensor_format *format,
+			   uint32_t raw,
+			   enum bt_mesh_sensor_value_status expected_status,
+			   float expected_f, bool exact)
+{
+	struct bt_mesh_sensor_value val = { .format = format };
+	enum bt_mesh_sensor_value_status status;
+	float f = 0.0f;
+
+	memcpy(val.raw, &raw, 4);
+	status = bt_mesh_sensor_value_to_float(&val, &f);
+	zassert_equal(status, expected_status);
+	if (expected_f == 0 || exact) {
+		zassert_equal(f, expected_f);
+	} else {
+		double error = (double)f / (double)expected_f;
+
+		zassert_between_inclusive(error, 0.99999, 1.00001);
+	}
+}
+
+static void check_from_special(const struct bt_mesh_sensor_format *format,
+			       enum bt_mesh_sensor_value_status special,
+			       int expected_err, uint32_t expected_raw)
+{
+	struct bt_mesh_sensor_value val = { 0 };
+	int err = bt_mesh_sensor_value_from_special_status(format, special, &val);
+
+	zassert_equal(err, expected_err);
+	if (expected_err == 0) {
+		zassert_mem_equal(val.raw, &expected_raw, 4);
+		zassert_equal(val.format, format);
+	}
+}
+
+static void check_to_status(const struct bt_mesh_sensor_format *format,
+			    uint32_t raw,
+			    enum bt_mesh_sensor_value_status expected_status)
+{
+	struct bt_mesh_sensor_value val = { .format = format };
+	enum bt_mesh_sensor_value_status status;
+
+	memcpy(val.raw, &raw, 4);
+	status = bt_mesh_sensor_value_get_status(&val);
+	zassert_equal(status, expected_status);
+}
+
+static uint32_t encoded(const struct format_spec *spec, double value)
+{
+	int64_t val = round(value / spec->ratio);
+
+	val &= GENMASK((spec->size * 8) - 1, 0);
+	return *(uint32_t *)&val;
+}
+
+/** Finds the biggest float less than or equal to @c limit. */
+static float biggest_float_leq(double limit)
+{
+	float val = limit;
+
+	return val <= limit ? val : nextafterf(val, -INFINITY);
```

Comment:
No, because converting a double to a float will round to the closest, which might be bigger than `limit` if rounding up. This checks if it was rounded up, and then picks the next possible float below if that's the case to ensure we have a value that is actually smaller than `limit`. Basically, it's a "convert double to float, force rounding down"-function.